File: leetcode/92-reverse-linked-list-II/main.go
package main

import (
	"fmt"
)

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, m int, n int) *ListNode {
	result := &ListNode{0, head}

	left := result
	for i := 0; i < m-1; i++ {
		left = left.Next
	}

	middle := left.Next
	right := middle.Next

	for i := 0; i < n-m; i++ {
		middle.Next = right.Next
		right.Next = left.Next
		left.Next = right
		right = middle.Next
	}

	return result.Next
}

func main() {
	list := &ListNode{
		Val: 1,
		Next: &ListNode{
			Val: 2,
			Next: &ListNode{
				Val: 3,
				Next: &ListNode{
					Val: 4,
					Next: &ListNode{
						Val: 5,
					},
				},
			},
		},
	}
	fmt.Println(reverseBetween(list, 2, 4))
}

File: leetcode/92-reverse-linked-list-II/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x, next=None):
        self.val = x
        self.next = next


class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        result = ListNode(0, head)
        left = result
        for i in range(m-1):
            left = left.next
        middle, right = left.next, left.next.next

        for i in range(n-m):
            middle.next = right.next
            right.next = left.next
            left.next = right
            right = middle.next

        return result.next


solution = Solution()
head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print(solution.reverseBetween(head, 2, 4))
